File: exercism/python/space-age/space_age.py
"""
Space Age
"""


class SpaceAge:
    """
    Class Space Age
    Given an age in seconds, calculate how old someone would be on
    our Planets
    """

    period = 31557600
    coef = {'earth': 1.0,
            'mercury': 0.2408467,
            'venus': 0.61519726,
            'mars': 1.8808158,
            'jupiter': 11.862615,
            'saturn': 29.447498,
            'uranus': 84.016846,
            'neptune': 164.79132}

    def __init__(self, seconds):
        """
        this would be a simple solution
            def on_earth(self):
                return float("{:.2f}".format(self.seconds / 31557600))
        """

        self.seconds = seconds

        for planet in SpaceAge.coef:
            self.add_method(planet)

    def add_method(self, planet):
        """
        dynamically creates the methods, based on the SpaceAge.coef index name.
        """

        setattr(self, "on_" + planet, lambda: self.sectoy(planet))

    def sectoy(self,  planet):
        """
        caltulates seconds to years
        """
        return round(self.seconds / (SpaceAge.coef[planet] * SpaceAge.period), 2)

File: exercism/go/space-age/space_age.go
package space

type Planet string

const period = 31557600.0

// Given an age in seconds, calculate how old someone would be on
// our Planets
var coef = map[Planet]float64{
	"Earth":   1.0,
	"Mercury": 0.2408467,
	"Venus":   0.61519726,
	"Mars":    1.8808158,
	"Jupiter": 11.862615,
	"Saturn":  29.447498,
	"Uranus":  84.016846,
	"Neptune": 164.79132}

func Age(seconds float64, planet Planet) float64 {

	// convert to lower case: Planet(strings.ToLower(string(planet)))
	if coefficient, OK := coef[planet]; OK {
		return seconds / (period * coefficient)
	}
	return -1
}
